cpp/src/arrow/python/numpy_to_arrow.cc

```diff
@@ -1069,7 +1033,7 @@ Status NumPyConverter::ConvertObjectsInfer() {
       return ConvertDateTimes();
     } else if (PyTime_Check(obj)) {
       return ConvertTimes();
-    } else if (PyObject_IsInstance(const_cast<PyObject*>(obj), Decimal.obj())) {
+    } else if (internal::PyDecimal_Check(obj)) {
```

Comment:
Aside: I feel like this function should be an addendum to jakevdp's "why is Python slow" blog post

Reply:
Part of the complication here is that we *could* do this by calling the import machinery once, but since we can't use RAII (and therefore `OwnedRef`s) we'd be leaking `PyObject*`s since we'd have to import them globally and then leave them there.

Reply:
That would remove the cost of importing, but would cause a memory leak.

Reply:
Alternatively, I can change this back to the way it was where the class held a reference to the `Decimal`  constructor.

Reply:
Actually that's what I'll do, I'm doing that in `builtin_convert.cc` and these should be consistent.

Reply:
This would also be faster if we had access to the Python C decimal object because we could just compare pointers (we'd lose the generality of `isinstance` though)